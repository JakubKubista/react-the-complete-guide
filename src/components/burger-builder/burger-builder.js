import React, { Component } from 'react';
import Aux from '../../hoc/aux';

class BurgerBuilder extends Component {
  render() {
    return (
      <Aux>
        <div>BurgerGraphics</div>
        <div>BuilderController</div>
      </Aux>
    );
  };
};

export default BurgerBuilder;