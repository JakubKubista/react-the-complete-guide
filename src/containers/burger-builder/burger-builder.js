/* eslint-disable no-unused-vars */
import React, { useEffect, useCallback } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { noop } from 'lodash';

import axios from '../../axios-service';
import Aux from '../../hoc/aux';
import withErrorHandler from '../../hoc/errorHandler';
import * as actions from '../../store/actions/index';
import { ROUTES } from '../../constants/routes';
import { MESSAGES } from '../../constants/labels';

import Burger from '../../components/burger-builder/burger/burger';
import BurgerControls from '../../components/burger-builder/burger-controls/burger-controls';
import Modal from '../../components/layout/modal/modal';
import OrderSummary from '../../components/burger-builder/order-summary/order-summary';
import Spinner from '../../components/layout/spinner/spinner';

const burgerBuilderStyle = {
  position: 'relative',
  top: '100px',
  textAlign: 'center'
};

export const BurgerBuilder = ({
  history,
  ingredients,
  price,
  purchasing,
  error,
  isSignedIn,
  onIngredientInit,
  onIngredientAdded,
  onIngredientRemoved,
  onPurchaseInit,
  onPurchasingOn,
  onPurchasingOff
}) => {

  useEffect(() => {
    onIngredientInit();
  }, [onIngredientInit]);

  const initBurger = useCallback(() => {
    let burger = null;

    if (error) {
      burger = <div style={burgerBuilderStyle}>
        {MESSAGES.ingredientsNotLoaded}
      </div>;
    } else {
      burger = <Spinner />;
    }

    return burger;
  }, [error]);

  const disabledIngredientsButton = useCallback(() => {
    const disabled = {
      ...ingredients
    }

    for (let key in disabled) {
      disabled[key] = disabled[key] <= 0;
    }

    return disabled;
  }, [ingredients]);

  const isPurchasable = useCallback(() => {
    const sum = (Object.values(ingredients))
      .reduce((a, b) => a + b, 0);
    return sum > 0;
  }, [ingredients]);

  const orderHandler = useCallback(() => {
    if (isSignedIn) {
      onPurchasingOn();
    } else {
      history.push(ROUTES.signIn);
    }
  }, [history, isSignedIn, onPurchasingOn])

  const purchasingContinueHandler = useCallback(() => {
    onPurchasingOff();
    onPurchaseInit();
    history.push(ROUTES.checkout);
  }, [history, onPurchaseInit, onPurchasingOff])

  let burger = initBurger();
  let orderSummary = null;

  if (ingredients) {
    burger = (
    <Aux>
      <Burger ingredients={ingredients} />

      <BurgerControls
        add={onIngredientAdded}
        remove={onIngredientRemoved}
        disabled={disabledIngredientsButton()}
        purchasable={isPurchasable()}
        purchasing={purchasing}
        price={price}
        order={orderHandler}
        isSignedIn={isSignedIn} />
    </Aux>);

    orderSummary = <OrderSummary
      ingredients={ingredients}
      price={price}
      cancel={onPurchasingOff}
      continue={purchasingContinueHandler} />
  }

  return (
    <Aux>
      <Modal
        show={purchasing}
        modalClosed={onPurchasingOff}>
        {orderSummary}
      </Modal>
      {burger}
    </Aux>
  );
};

BurgerBuilder.propTypes = {
  history: PropTypes.shape({
    push: PropTypes.func
  }),
  ingredients: PropTypes.shape({
    salad: PropTypes.number,
    cheese: PropTypes.number,
    meat: PropTypes.number,
    bacon: PropTypes.number
  }),
  price: PropTypes.number,
  purchasing: PropTypes.bool,
  error: PropTypes.bool,
  isSignedIn: PropTypes.bool,
  onIngredientInit: PropTypes.func,
  onIngredientAdded: PropTypes.func,
  onIngredientRemoved: PropTypes.func,
  onPurchaseInit: PropTypes.func,
  onPurchasingOn: PropTypes.func,
  onPurchasingOff: PropTypes.func,
};

BurgerBuilder.defaultProps = {
  history: {
    push: noop
  },
  ingredients: {},
  price: 0,
  purchasing: false,
  error: null,
  isSignedIn: false,
  onIngredientAdded: noop,
  onIngredientRemoved: noop,
  onPurchaseInit: noop,
  onPurchasingOn: noop,
  onPurchasingOff: noop
};

const mapStateToProps = state => {
  return {
    ingredients: state.burgerBuilder.ingredients,
    price: state.burgerBuilder.price,
    purchasing: state.burgerBuilder.purchasing,
    error: state.burgerBuilder.error,
    isSignedIn: state.auth && state.auth.token !== null
  };
};

const mapDispatchToProps = dispatch => {
  return {
    onIngredientInit: () => dispatch(actions.ingredientInit()),
    onIngredientAdded: (name) => dispatch(actions.ingredientAdd(name)),
    onIngredientRemoved: (name) => dispatch(actions.ingredientRemove(name)),
    onPurchaseInit: () => dispatch(actions.purchaseInit()),
    onPurchasingOn: () => dispatch(actions.purchasingOn()),
    onPurchasingOff: () => dispatch(actions.purchasingOff())
  }
};

const BurgerBuilderWithErrorHandler = withErrorHandler(BurgerBuilder, axios)

export default connect(mapStateToProps, mapDispatchToProps)(
  BurgerBuilderWithErrorHandler
);