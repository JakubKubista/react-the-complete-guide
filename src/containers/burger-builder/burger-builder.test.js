import React from 'react';
import {create} from 'react-test-renderer';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { BrowserRouter } from 'react-router-dom';

import { BurgerBuilder } from './burger-builder';
import BurgerControls from '../../components/burger-builder/burger-controls/burger-controls';

configure({adapter: new Adapter()});

describe('<BurgerBuilder/>', () => {
  let wrapper = null;

  beforeEach(() => {
    wrapper = shallow(<BurgerBuilder onIngredientInit={() => {}} />);
    wrapper.setProps({ingredients: {salad: 0}});
  })

  it('should match to snapshot', () => {
    const renderer = create(
      <BrowserRouter>
        <BurgerBuilder onIngredientInit={() => {}} />
      </BrowserRouter>
    );
    expect(renderer.toJSON()).toMatchSnapshot();
  });

  it('should show render <BurgerControls /> if ingredients are set', () => {
    expect(wrapper.find(BurgerControls)).toHaveLength(1);
  });

  it('should have correct ingredients state', () => {
    const state = {ingredients: {salad: 4}};

    wrapper.setState(state);

    expect(wrapper.state()).toEqual(state);
  });
});