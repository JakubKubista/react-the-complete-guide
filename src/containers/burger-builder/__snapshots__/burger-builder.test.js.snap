// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BurgerBuilder/> should match to snapshot 1`] = `
Array [
  <div
    style={
      Object {
        "opacity": "0",
        "transform": "translateY(-100vh)",
      }
    }
  >
    <h3>
      Your Order
    </h3>
    <p>
      A delicious burger with the following ingredients:
    </p>
    <ul />
    <p>
      <strong>
        Total Price
        : 
        0.00
      </strong>
    </p>
    <p>
      Continue to Checkout?
    </p>
    <button
      className=" "
      onClick={[Function]}
    >
      CANCEL
    </button>
    <button
      className=" "
      onClick={[Function]}
    >
      CONTINUE
    </button>
  </div>,
  <div>
    <div>
      <div />
      <div />
    </div>
    <p>
      Please start adding ingredients.
    </p>
    <div />
  </div>,
  <div>
    <p>
      Price
      : 
      <b>
        USD
         
        0.00
      </b>
    </p>
    <div>
      <div>
        Salad
      </div>
      <button
        onClick={[Function]}
      >
        Less
      </button>
      <button
        onClick={[Function]}
      >
        More
      </button>
    </div>
    <div>
      <div>
        Cheese
      </div>
      <button
        onClick={[Function]}
      >
        Less
      </button>
      <button
        onClick={[Function]}
      >
        More
      </button>
    </div>
    <div>
      <div>
        Meat
      </div>
      <button
        onClick={[Function]}
      >
        Less
      </button>
      <button
        onClick={[Function]}
      >
        More
      </button>
    </div>
    <div>
      <div>
        Bacon
      </div>
      <button
        onClick={[Function]}
      >
        Less
      </button>
      <button
        onClick={[Function]}
      >
        More
      </button>
    </div>
    <button
      disabled={true}
      onClick={[Function]}
    >
      SIGN IN TO ORDER
    </button>
  </div>,
]
`;
